### ISSUE TITLE
Update all Demos to Python 3

### ISSUE THREAD

**Issue by pyscripter** [pyscripter/python4delphi#150]
Python 2 has reached the end of its life, and we can assume that most of the users will have installed python 3.   Most of the demos use python 2 print statements and some may use other python 2 idioms.

To facilitate experimentation and learning with the demos they should be converted to python 3.

*Also all forms should be changed to text format and the old application icon replaced with the default icon.*

**Comment by drbond** [comment]
Agree, especially as the only tutorial seems to be the demos. That isn't the only issue with several of the demos. They appear to have been written for a very old version of Delphi so don't compile in current versions of Delphi because of , for instance, a missing icon.

**Comment by pyscripter** [comment]
- [x] Demo 1
- [ ] Demo 2
- [ ] Demo 3
- [ ] Demo 4
- [ ] Demo 5
- [ ] Demo 6
- [ ] Demo 7
- [ ] Demo 8
- [ ] Demo 9
- [ ] Demo 10
- [ ] Demo 12
- [ ] Demo 13
- [ ] Demo 14
- [ ] Demo 15
- [ ] Demo 16
- [ ] Demo 17
- [ ] Demo 18
- [ ] Demo 19
- [ ] Demo 20
- [ ] Demo 22
- [ ] Demo 23
- [ ] Demo 24
- [ ] Demo 25
- [ ] Demo 26
- [ ] Demo 27
- [ ] Demo 28
- [ ] Demo 29
- [ ] Demo 30
- [ ] Demo 31
- [ ] Demo 32
- [X] Demo 33
- [ ] Demo 34

- [ ] Demo 21
- [ ] Demo 21
- [ ] Demo 21

**Comment by JT-ITStation** [comment]
> 
> 
> Agree, especially as the only tutorial seems to be the demos. That isn't the only issue with several of the demos. They appear to have been written for a very old version of Delphi so don't compile in current versions of Delphi because of , for instance, a missing icon.

try to open project option and set defaut to icon program

Demo2 work

Demo 3 KO : exeption raise execption :
 s_value 'Missing parentheses in call to ''print''. Did you mean print("Python Dll: ", sys.version)? (<string>, line 2)'

**Comment by pyscripter** [comment]
> s_value 'Missing parentheses in call to ''print''. Did you mean print("Python Dll: ", sys.version)? (, line 2)'

In python 3 print is a function:

eg. Python 2
print "Python Dll: ", sys.version
need to be converted to
print("Python Dll: ", sys.version)

This is what this issue is about.

**Comment by ReinierNL** [comment]
Hi Kiriakos, I see you are working on the demos right now.
I wanted to start also now, but not get in your way. Can we divide some tasks?
If you are working from low (3) to high, then let me (unless advised otherwise) start from high (34) to low.

**Comment by pyscripter** [comment]
Hi Renier,  just did Demo 3 to understand what needs to be changed.
I am planning  to do anything else right now.  So just feel free to work your way up.

-  Don't forget to change the code in the InitScript property
- Get rid of the old icons
- Make the default platform Win64 (I recommend users to install python 64bits)
- Change the fonts in the memos to Consolas size 10.

- In Demo 3 I had to change the Form.Position to Default, so that it looks the right size in a high-dpi monitor.
- Also added the Enable Run Time themes option.

**Comment by ReinierNL** [comment]
Excellent. Thanks for the instructions :-)
Not sure if I know what you mean by the first bullet though; so far I did not encounter any problems that involved InitScript.
But I'll have a look at it anyway.
Also, about the Form.Position, I'll try to think of it; the thing is I don't have a high dpi monitor here so I won't notice it if something is off there.

**Comment by ReinierNL** [comment]
And about Demo34: that one is already Python 3 compatible. So you may check the box there.

**Comment by ReinierNL** [comment]
And Demo02 too. I just converted the dfm to txt in the PR I just submitted.

**Comment by pyscripter** [comment]
> Not sure if I know what you mean by the first bullet though; so far I did not encounter any problems that involved InitScript.

PythonEngine has a property called InitScript.  In some demos it contains code with python 2 print statements

**Comment by pyscripter** [comment]
> Also, about the Form.Position, I'll try to think of it; the thing is I don't have a high dpi monitor here so I won't notice it if something is off there.

Just set the Form.Postion property to Default.

**Comment by ReinierNL** [comment]
Oh. Well, in the ones I did today I've modified it to Default. From here on I'll leave it as is.

**Comment by ReinierNL** [comment]
Hi Kiriakos,  I just took a look at Demo13.  That one may probably also be moved to 'deprecated', because it uses TPythonDatabase, which is declared in unit PythonDatabase.pas which is under the Unsupported folder.
And the demo also uses TTable which is no longer available in Delphi (I think since Delphi 10.2; see for example https://stackoverflow.com/questions/44314923/cannot-resolve-unit-name-dbtables-at-line-20-bde).

**Comment by pyscripter** [comment]
@ReinierNL 
Done.   

The next couple ones are reported as Kylix ready.  We can remove Kylix related stuff.

**Comment by ReinierNL** [comment]
About demos 18, 19 and 20:  apparently those are demos for C++ Builder, which I don't have.
What I *could* do with these, is checking the syntax of the Python code.
But I am postponing that for the time being.

**Comment by ReinierNL** [comment]
About demos 22 and 23:
They are a bit different from the other demos in the sense that they don't have a Memo2 for the output, but when you run them, a "Python console" appears. However, I don't see any output appearing anywhere.
Furthermore, the Python code uses *threading._test()*, which apparently does not exist (anymore?), and I don't know what to replace it with.
Perhaps we should make a separate issue for these 2 demos?

**Comment by pyscripter** [comment]
> About demos 18, 19 and 20

I have created a new issue regarding C++ Builder Support

**Comment by pyscripter** [comment]
> About demos 22 and 23:

These are quite old demos (python 1.6 time!)

Presumably a threading_test() method needs to be provided by the user.  Something like:

```python
def thread_function(name):
    import time
    print("Thread %s: starting" % name)
    time.sleep(1)
    print("Thread %s: finishing" % name)

def threading_test():
    t = threading.Thread(target=thread_function, args=("Test thread",))
    t.start()
    t.join()
```

Leave them on the side for the time being and I will get back to them.

**Comment by JT-ITStation** [comment]
I Can check for c++ builder... If it compile, i have missed your
présentatio using Delphi and python, IS their any Replay ?

**Comment by pyscripter** [comment]
> IS their any Replay ?

https://github.com/pyscripter/python4delphi/tree/master/Tutorials/Webinar%20I

**Comment by pyscripter** [comment]
@ReinierNL What is the state of work on this issue?

**Comment by ReinierNL** [comment]
Oh, I am sorry I didn't see your message earlier!

I think you can check the boxes for demos 16, 17, 21, 27, 28, 30 and 32.
All the other demos except Demo26 have also been updated with respect to the Python 3 syntax.
Apparently I somehow forgot to update demo 26. I will still do that one.
The other demos run, but there are issues:
Demos 22 end 23:  Project Demo22.exe raised exception class EPyAttributeError with message 'AttributeError: module 'threading' has no attribute '_test''.
Demo 25:  runs, but tests fail for: Sequences, Mappings, Objects
Demo 29:  runs, but:   ImportError: DLL load failed: The specified module could not be found. (something to do with installation of PIL I guess)

I did not look at the FMX demo or the FPC demos.
I am not sure if I will have time to do that before Wednesday. You'll know if you get PRs.

**Comment by ReinierNL** [comment]
Oh, and demo 24 does not exist anymore.

**Comment by pyscripter** [comment]
@ReinierNL Thank you very much.

Under Unsupported/Old Docs there is a file 
https://github.com/pyscripter/python4delphi/blob/master/Unsupported/Old%20docs/Tutorial.txt
that provided detailed descriptions of each demo.  
Would it be possible to have a look and decide whether an edited/amended version of this should be moved back to the Demos folder?

**Comment by ReinierNL** [comment]
@pyscripter: 
I am hesitating.
It looks like quite a lot of work and although I think it's a really nice tutorial, the added value relative to trying out the demos and looking at their source code is somewhat limited.
It also depends on when you would like to have it ready (before the Delphi webinar on Wednesday?) and what other improvements you'd like to have made by then.
Personally, I'd like to first have a look at the other demos (FMX; FPC), then have a look at the demos that don't work 100% well before I start working on that tutorial.

What we could also do is start by adding it back, but include some extra text at the beginning, saying that the tutorial itself is not up-to-date with respect to the Python code (using Python 2 syntax) and that we will soon take care of that and that help is appreciated.

What do you think?

**Comment by pyscripter** [comment]
@ReinierNL Thanks for the feedback.  Your suggested work plan sounds great.  Please don't feel any pressure to do stuff.  Do what you like, as much as you like and when you have time available.  Your contributions are much appreciated.

**Comment by drbond** [comment]
Just my two penny’s worth: When I came to trying out Python4Delphi just over a week ago after watching the Webinar, I couldn’t make much sense of the demos + source code. Demo 1 didn’t work out of the box initially  but the reason was fairly obvious: pre-version 3 Python. Several of the other demos didn’t work either, e.g. TTable not supported any longer. But the demos have been updated so maybe this isn’t the issue it was. However, I experienced a couple of other issues which haven’t necessarily gone away for a newcomer and which I overcame after a request for help - see thread I started. I was very fortunate to receive excellent support from a member of the team which got me going. The value of a good tutorial is that it provides orientation to get you in to the tool. It also avoids you making errors early on which can deter further interest. I couldn’t have written the Chapter80Sample pdf from the demos and source code alone. There are still some aspects of the image processing source code that I don’t understand, e.g. the checkbox and its use in an If Then Else statement. I also found that this code fell over if images were bigger than about a Megabyte but I haven’t found time yet to test whether it is caused by Python4Delphi code or a limitation in the Python. An explantation/some guidance on this aspect would be helpful. I didn’t find the demos or the  source code sufficient to gain sufficient understanding to be confident to write the sample chapter. Instead, I relied on other tutorials (circa 2001) available from the Web and, of course, Webinar 1. To anyone well versed in the source code or associated with its development, it may be obvious how to use Python4Delphi but for a newbie like me it was not. However, I appreciate that time is pressing but any tutorial is better than none. 

Looking forward to Webinar 2. I really enjoyed Webinar 1 and did get a lot from it - it sold Python4Delphi to me!

Best wishes

Kevin

**Comment by ReinierNL** [comment]
@drbond:  Well, Kevin,  thanks for adding those two pennies. That may help me overcome my hesitations :-)
Perhaps I'll have a go at it tomorrow.

In the mean time, let me report on what I found concerning the FMX and FPC examples:
FMX:   I could verify this only on Windows. There it works without problems. There are no Python 2 issues. 
But the real proof of the pudding would be to run it on OSX, which I don't have access to.

FPC: I've downloaded and installed Lazarus and then opened the first example. It requires the packages to be installed. Fair enough. So I opened the package file and tried to compile.
It turns out that lines 1679 and 1681 of VarPyth won't compile because of the Variant to NativeInt cast/assignment that takes place in the function calls of PyList_SetItem and PyTuple_SetItem. Apparently FreePascal does not handle assignments of Variants to other types very well.
I tried it without variants (passing NativeInt(@AKey) in stead of Variant(AKey) ) and actually the packages compiled with this modification.
However, the example (FPC\Demo01) now gives an error upon startup, which I can't explain or fix.
So I am a bit stuck there.

@pyscripter: shall we (shall I?) create separate issues for those demos that don't work well yet?
Then this issue (150) could be closed. ISTM that this thread is becoming a bit long.

**Comment by pyscripter** [comment]
@ReinierNL 
The FMX demo should work in OSX *and Linux* with minimal modifications required to find the right version of python.   I did try that a few months ago.

The Variants unit used to compile in FPC last time I tried it (it was a while ago). I may have introduced some incompatibilities.  See also the FPC notes in https://github.com/pyscripter/python4delphi/wiki/SupportedPlatforms

Let's close this and open a few new issues:
- FPC support (Variants and Demo)
- Complete demos 22/23
- Update and bring back the Tutorial.txt

Anything else?

David Intersimone has produced a C++Builder tutorial and probably will be updating the C++Demos.
